File: apps/mobile/lib/core/network/interceptors/auth_interceptor.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod_clean_architecture/core/storage/auth_token_storage.dart';

/// Lista de endpoints publicos do backend Augustus que **nao** devem receber
/// `Authorization: Bearer`.
const _publicAuthPaths = <String>{
  '/auth/login',
  '/auth/register',
  '/auth/refresh',
  '/auth/verify-email',
  '/auth/resend-verification',
};

bool _isPublicAuthPath(String path) {
  return _publicAuthPaths.any(path.contains);
}

/// Injeta `Authorization: Bearer <accessToken>` em toda request, exceto
/// para endpoints publicos de auth. Le do storage de forma assincrona.
class AuthInterceptor extends Interceptor {
  AuthInterceptor({required AuthTokenStorage tokenStorage})
      : _tokenStorage = tokenStorage;

  final AuthTokenStorage _tokenStorage;

  @override
  Future<void> onRequest(
    RequestOptions options,
    RequestInterceptorHandler handler,
  ) async {
    if (_isPublicAuthPath(options.path)) {
      return handler.next(options);
    }
    final token = await _tokenStorage.readAccess();
    if (token != null && token.isNotEmpty) {
      options.headers['Authorization'] = 'Bearer $token';
    }
    return handler.next(options);
  }
}

File: apps/web/src/app/core/auth/auth.interceptor.ts
import { HttpInterceptorFn } from '@angular/common/http';
import { inject } from '@angular/core';
import { catchError, from, switchMap, throwError } from 'rxjs';
import { AuthService } from './auth.service';
import { TokenStorage } from './token-storage';

const SKIP_PATHS = [
  '/api/auth/login',
  '/api/auth/register',
  '/api/auth/refresh',
  '/api/auth/verify-email',
  '/api/auth/resend-verification',
];

function isSkip(url: string): boolean {
  return SKIP_PATHS.some((p) => url.includes(p));
}

export const authInterceptor: HttpInterceptorFn = (req, next) => {
  const tokenStorage = inject(TokenStorage);
  const auth = inject(AuthService);

  let request = req;
  if (!isSkip(req.url)) {
    const token = tokenStorage.lerAccess();
    if (token) {
      request = req.clone({
        setHeaders: { Authorization: `Bearer ${token}` },
      });
    }
  }

  return next(request).pipe(
    catchError((err) => {
      if (err?.status !== 401 || isSkip(req.url)) {
        return throwError(() => err);
      }
      return from(auth.refresh()).pipe(
        switchMap((ok) => {
          if (!ok) {
            return throwError(() => err);
          }
          const novoToken = tokenStorage.lerAccess();
          if (!novoToken) {
            return throwError(() => err);
          }
          const retried = req.clone({
            setHeaders: { Authorization: `Bearer ${novoToken}` },
          });
          return next(retried);
        }),
      );
    }),
  );
};
